wallet: track mempool conflicts with wallet transactions

ishaanam opened the pull request:
The `mempool_conflicts` variable is added to `CWalletTx`, it is a set of txids of txs in the mempool conflicting with the wallet tx or a wallet tx's parent. This PR only changes how mempool-conflicted txs are dealt with in memory.

`IsSpent` now returns false for an output being spent by a mempool conflicted transaction where it previously returned true.

A txid is added to `mempool_conflicts` during  `transactionAddedToMempool`. A txid is removed from `mempool_conflicts` during  `transactionRemovedFromMempool`.

This PR also adds a `mempoolconflicts` field to the `gettransaction` wallet RPC result.

Builds on #27145
Second attempt at #18600

ishaanam committed:
wallet refactor: use CWalletTx member functions to determine tx state

ishaanam committed:
wallet: track mempool conflicts

Behavior changes are:
- if a tx has a mempool conflict, the wallet will not attempt to
  rebroadcast it
- if a txo is spent by a mempool-conflicted tx, that txo is no
  longer considered spent

evansmj commented on src/wallet/transaction.h:
```diff
@@ -315,7 +321,9 @@ class CWalletTx
-    bool isConflicted() const { return state<TxStateConflicted>(); }
+    bool isConflicted() const { return state<TxStateConflicted>() || state<TxStateMempoolConflicted>(); }
+    bool isMempoolConflicted() const { return state<TxStateMempoolConflicted>(); }
+    bool isBlockConflicted() const { return state<TxStateConflicted>(); }
```
What do you think of renaming the `TxStateConflicted` struct to `TxStateBlockConflicted`, since that one is used for block conflicts and there is now a difference here between block and mempool conflicts?

`//! State of rejected transaction that conflicts with a confirmed block.
struct TxStateConflicted {
...
};`

ishaanam committed:
scripted-diff: wallet: s/TxStateConflicted/TxStateBlockConflicted

-BEGIN VERIFY SCRIPT-
sed -i 's/TxStateConflicted/TxStateBlockConflicted/g' src/wallet/wallet.cpp src/wallet/interfaces.cpp src/wallet/transaction.h src/wallet/transaction.cpp
sed -i 's/isConflicted/isBlockConflicted/g' src/wallet/transaction.h src/wallet/wallet.cpp
-END VERIFY SCRIPT-

ishaanam committed:
wallet, rpc: show mempool conflicts in `gettransaction` result

ishaanam commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +129,148 @@ def run_test(self):
+        self.disconnect_nodes(0, 1)
+
+        # alice has all 0 txs, bob has 3
+        assert_equal(len(alice.getrawmempool()), 0)
```
Fixed

ishaanam commented on src/wallet/wallet.h:
```diff
@@ -337,6 +337,9 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     void AddToSpends(const CWalletTx& wtx, WalletBatch* batch = nullptr) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
 
+    // Map of wtx hash to a set of mempool conflict hashes
+    std::map<uint256, std::set<uint256>> MempoolConflicts GUARDED_BY(cs_wallet);
```
I agree it is better to track conflicts on a transaction level. I've implemented it this way now and it looks much better.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1482,8 +1545,8 @@ void CWallet::blockConnected(ChainstateRole role, const interfaces::BlockInfo& b
 
     // Scan block
     for (size_t index = 0; index < block.data->vtx.size(); index++) {
-        SyncTransaction(block.data->vtx[index], TxStateConfirmed{block.hash, block.height, static_cast<int>(index)});
         transactionRemovedFromMempool(block.data->vtx[index], MemPoolRemovalReason::BLOCK);
+        SyncTransaction(block.data->vtx[index], TxStateConfirmed{block.hash, block.height, static_cast<int>(index)});
```
In 28a7e883c2f9249213fddbe24948ec5bb90b0fad "wallet: track mempool conflicts"

It's not clear to me that this change has any effect.

ishaanam replied:
I think before this was to fix a bug, but I fixed that bug in a different way. Regardless, doesn't it make more sense to run `transactionRemovedFromMempool` before syncing the transaction as confirmed?

glozow replied:
fwiw, it seems better to keep it the way it was if it's not doing anything. What was the bug?

ishaanam replied:
I don't remember exactly what the bug was, but I know that it had something to do with assuming that all transactions that we call `transactionRemovedFromMempool` on are `TxStateInMempool`. That being said, I will change it back to the way it was before because it is not causing any problems right now.

achow101 replied:
I think this hasn't been reverted yet?

ishaanam replied:
Should be fixed now.

achow101 commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +129,148 @@ def run_test(self):
+
+        self.disconnect_nodes(0, 1)
+
+        raw_tx = alice.createrawtransaction(inputs=[unspents[0]], outputs=[{bob.getnewaddress() : 24.99999}])
+        raw_tx1 = alice.signrawtransactionwithwallet(raw_tx)['hex']
```
In 003efbbe45079c4416810a025b2bc372559dff15 "test: Add tests for wallet mempool conflicts"

This can be compressed into one line using the `send` RPC.

```suggestion
        raw_tx1 = alice.send(outputs=[{bob.getnewaddress(): 24.9999}], inputs=[unspents[0]], add_to_wallet=False)["hex"]
```

ishaanam replied:
These two things are not technically equivalent, because `send` creates a change output here, which results in intermittent failures for the rest of the test. Is there any way to tell `send` not to create a change output?

achow101 commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +129,148 @@ def run_test(self):
+        # we will be disconnecting this block in the future
+        tx2_conflict = alice.sendrawtransaction(tx2_conflict)
+        assert_equal(len(alice.getrawmempool()), 1)
+        blk = self.generate(self.nodes[0], 11, sync_fun=self.no_op)[0]
```
In 003efbbe45079c4416810a025b2bc372559dff15 "test: Add tests for wallet mempool conflicts"

It would be useful to have a comment that explains that 11 blocks are mined so that when they are invalidated, `tx2_conflict` does not get put back into the mempool.

ishaanam replied:
I've added a comment.

achow101 commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +129,148 @@ def run_test(self):
+
+        self.log.info("Test a scenario where a transaction has a mempool conflict")
+
+        unspents = [{"txid" : element["txid"], "vout" : element["vout"]} for element in alice.listunspent()]
```
In 003efbbe45079c4416810a025b2bc372559dff15 "test: Add tests for wallet mempool conflicts"

This isn't actually necessary since `listunspent` has these fields already and the extra ones are just ignored.

ishaanam replied:
Fixed

murchandamus commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +129,148 @@ def run_test(self):
+        raw_tx = alice.createrawtransaction(inputs=[unspents[2]], outputs=[{bob.getnewaddress() : 24.9899}])
+        tx3 = alice.signrawtransactionwithwallet(raw_tx)['hex']
+
+        tx1 = alice.sendrawtransaction(tx1)
```
In 003efbbe45079c4416810a025b2bc372559dff15 "test: Add tests for wallet mempool conflicts":
It’s a bit confusing that you take an input `tx1` here which is the 'hex' string of the raw transaction, but then assign the resulting 'txid' back to `tx1`.

ishaanam replied:
I've replaced `tx1` with `tx1_txid` here and for all other instances of this.

murchandamus commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +129,148 @@ def run_test(self):
+
+        assert_equal(alice.gettransaction(tx1)["mempoolconflicts"], [])
+
+        # now all of alice's outputs should be considered spent
```
In 003efbbe45079c4416810a025b2bc372559dff15 "test: Add tests for wallet mempool conflicts":
Since `tx2` replaced `tx1`, why would `unspents[0]` be considered spent here?

ishaanam replied:
Because `tx3` replaced `tx2`, so `tx1` is now `TxStateInactive` instead of `TxStateMempoolConflicted`, making `unspents[0]` spent.

ishaanam replied:
I have added a comment before all of the tests in this file explaining how these tests create conflicts.

murchandamus commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +129,148 @@ def run_test(self):
+        self.generate(self.nodes[0], 3)
+        assert_equal(alice.getbalance(), 75)
+
+    def test_mempool_and_block_conflicts(self):
+        alice = self.nodes[0].get_wallet_rpc("alice")
+        bob = self.nodes[1].get_wallet_rpc("bob")
```
Yeah, this could be a problem if you only run one of the tests separately instead of the whole suite. Then that test would fail, since it’s setup depends on the prior test having been run before.

ishaanam replied:
Done

murchandamus commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +129,148 @@ def run_test(self):
+        raw_tx3 = bob.signrawtransactionwithwallet(raw_tx)['hex']
+        tx3 = bob.sendrawtransaction(raw_tx3)
+
+        # alice has 0 txs, bob has 3
```
In 003efbbe45079c4416810a025b2bc372559dff15 "test: Add tests for wallet mempool conflicts":
I’m having trouble keeping track of all the things going on here. Perhaps you could add in these comments which transactions each node now knows about

ishaanam replied:
I agree that it gets a bit confusing, so I've added a lot more comments.

murchandamus commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +129,148 @@ def run_test(self):
+        # bob broadcasts tx_1 conflict
+        tx1_conflict = bob.sendrawtransaction(tx1_conflict)
+        assert_equal(len(alice.getrawmempool()), 0)
+        assert_equal(len(bob.getrawmempool()), 2)
```
In 003efbbe45079c4416810a025b2bc372559dff15 "test: Add tests for wallet mempool conflicts":
`tx1_conflict` should displace `tx1`, but why does Bob not have 3 again now? Shouldn’t Bob know about `tx1`, `tx2`, and `tx3` before, and then `tx1_conflict`, `tx2`, and `tx3` after?

ishaanam replied:
`tx1_conflict` also displaces `tx3` because `tx3` spends `tx1`, so it is also considered mempool-conflicted. So instead `tx1_conflict` and `tx2` are the only transactions left in the mempool. I've added a comment indicating this.

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -752,8 +752,7 @@ bool CWallet::IsSpent(const COutPoint& outpoint) const
         if (mit != mapWallet.end()) {
-            int depth = GetTxDepthInMainChain(mit->second);
-            if (depth > 0  || (depth == 0 && !mit->second.isAbandoned()))
+            if (mit->second.isConfirmed() || mit->second.InMempool() || (mit->second.isInactive() && !mit->second.isAbandoned()))
```
`depth > 0  || (depth == 0 && !mit->second.isAbandoned())` is saying that the spending transaction should be confirmed, or: must not be (block) conflicted, and it must not be abandoned, for the corresponding output to be considered spent. 

`mit->second.isConfirmed() || mit->second.InMempool() || (mit->second.isInactive() && !mit->second.isAbandoned())`  is saying that the spending transaction should be confirmed, or in the mempool, or: must be inactive and not abandoned for the corresponding output to be considered spent. 

In this context, saying that a transaction must not be conflicted is logically equivalent to saying that the transaction must be inactive and not abandoned. I will add a comment to clarify this.

ishaanam commented on src/wallet/interfaces.cpp:
```diff
@@ -91,7 +91,7 @@ WalletTxStatus MakeWalletTxStatus(const CWallet& wallet, const CWalletTx& wtx)
     WalletTxStatus result;
     result.block_height =
         wtx.state<TxStateConfirmed>() ? wtx.state<TxStateConfirmed>()->confirmed_block_height :
-        wtx.state<TxStateConflicted>() ? wtx.state<TxStateConflicted>()->conflicting_block_height :
```
I think that `TxStateBlockConflicted` and `TxStateMempoolConflicted` are a bit more clear, because with something like `TxStateUnconfirmedConflict` instead of `TxStateMempoolConflicted`, it is unclear whether the conflict is in the mempool, or just hasn't been broadcasted yet. Additionally, I felt that it was more uniform with the other transaction states to be describing the state of the transaction instead of the state of a conflict. Eg. `TxStateBlockConflicted` indicates that the transaction is block-conflicted, and similarly `TxStateConfirmed` indicates that the transaction is confirmed. On the other hand, `TxStateConfirmedConflict` indicates the state of the conflict. 

That being said, if you or others think that `TxStateConfirmedConflict` or something similar is better, I would be open to changing it.

murchandamus replied:
I don’t feel strongly about it, please feel free to mark this comment as resolved.

ishaanam commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +129,148 @@ def run_test(self):
+        alice = self.nodes[0].get_wallet_rpc("alice")
+
+        self.nodes[1].createwallet("bob")
+        bob = self.nodes[1].get_wallet_rpc("bob")
```
Done

ishaanam commented on test/functional/wallet_abandonconflict.py:
```diff
@@ -232,7 +232,9 @@ def run_test(self):
         # Invalidate the block with the double spend. B & C's 10 BTC outputs should no longer be available
-        self.nodes[0].invalidateblock(self.nodes[0].getbestblockhash())
+        blk = self.nodes[0].getbestblockhash()
+        self.generate(self.nodes[1], 10)
```
I've added this comment explaining why 10 blocks need to be mined:
```
# mine 10 blocks so that when the blk is invalidated, the transactions are not
# returned to the mempool
```

ishaanam commented:
Rebased and addressed most of the review comments (thanks @achow101 and @murchandamus). I will address the remaining comments soon.

ishaanam commented:
I have addressed the remaining questions, the only changes are the addition of comments to the functional test.

murchandamus reviewed (commented):
ACK  22778ac98f1bcb695de8ccf0b02fe9ebe39b01aa

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -752,8 +752,9 @@ bool CWallet::IsSpent(const COutPoint& outpoint) const
         if (mit != mapWallet.end()) {
-            int depth = GetTxDepthInMainChain(mit->second);
-            if (depth > 0  || (depth == 0 && !mit->second.isAbandoned()))
+            // A transaction is considered spent if a spending transaction
```
Fixed

usuarrio-id81214293 replied:
> Fixed

Need enter wallet

murchandamus commented:
ACK https://github.com/bitcoin/bitcoin/pull/27307/commits/d4259cdaff2128c25e877d6a64e6ce15ecb9beca, only documentation change

murchandamus commented on src/wallet/transaction.h:
```diff
@@ -48,7 +48,12 @@ struct TxStateBlockConflicted {
+
+//! State of transaction with one or more mempool conflicts.
+struct TxStateMempoolConflicted {
+    std::string toString() const { return strprintf("MempoolConflicted"); }
```
In  a6ae5b23b1497ab6f4899db49348db623700a2d8 "wallet: track mempool conflicts":

Nit: Should this perhaps also inform about the conflicting transactions like the _BlockConflicted_ state informs about the block?

ishaanam replied:
This could be done in a follow-up if people want it, because currently the conflicting transactions are stored in the `CWalletTx`, and not the transaction state, so this function is not aware of the txids of the mempool conflicts.

murchandamus reviewed (commented):
ACK 71059cf4267d80d2934b85a7046fc7b64900378b

glozow commented on src/wallet/wallet.cpp:
```diff
@@ -1419,6 +1419,36 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+
+        // For each wallet transaction spending this prevout...
+        for (TxSpends::const_iterator _it = range.first; _it != range.second; ++_it) {
+            // if this tx is the same as the tx which was just added to the mempool, continue
+            if (_it->second == conflict_hash) continue;
```
in 067189eddee1a2209c6c42741a91601c5514e170

Got a little nervous seeing the txid comparison and "same tx." Is it possible to run into a same-txid-different-witness situation here?

ishaanam replied:
Good catch, I think that is possible. Would just using wtxids everywhere here fix this?

glozow replied:
Yes I'd expect so

ishaanam replied:
Thanks, that should be fixed now

achow101 replied:
If we get the same txid but different wtxid, I'm not sure that the result is necessarily one that makes sense. Most things in the wallet are keyed and listed by txid, so with this change, we'd end up with a transaction that's listed as conflicted by it's own txid. Also, we won't ever update the wallet to store a transaction with a different wtxid.

ishaanam replied:
> If we get the same txid but different wtxid, I'm not sure that the result is necessarily one that makes sense. Most things in the wallet are keyed and listed by txid, so with this change, we'd end up with a transaction that's listed as conflicted by it's own txid. Also, we won't ever update the wallet to store a transaction with a different wtxid.

Sorry I didn't see this earlier. After looking at this again I don't think that same-txid-different-witness will be a problem here. Please see: https://github.com/bitcoin/bitcoin/pull/27307#discussion_r1484797295

glozow commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -417,6 +421,10 @@ static std::vector<RPCResult> TransactionDescriptionString()
            }},
            {RPCResult::Type::STR_HEX, "replaced_by_txid", /*optional=*/true, "Only if 'category' is 'send'. The txid if this tx was replaced."},
            {RPCResult::Type::STR_HEX, "replaces_txid", /*optional=*/true, "Only if 'category' is 'send'. The txid if this tx replaces another."},
+           {RPCResult::Type::ARR, "mempoolconflicts", "Conflicting transaction currently in the mempool.",
```
c71736d372c65b0978671501029e202574c6bbef

Also good to document - is it just directly conflicting, or descendants as well?

ishaanam replied:
Done

glozow commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +130,165 @@ def run_test(self):
+        self.log.info("Test a scenario where a transaction has a mempool conflict")
+
+        unspents = alice.listunspent()
+        assert_equal(len(unspents), 3)
```
71059cf4267d80d2934b85a7046fc7b64900378b

There's not anything preventing us from checking the exact contents of `listunspent` right? Here and in the rest of the test.

ishaanam replied:
Good point, the tests now check the exact contents.

glozow commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +130,165 @@ def run_test(self):
+
+        alice.sendrawtransaction(tx3)
+
+        assert_equal(alice.gettransaction(tx1_txid)["mempoolconflicts"], [])
```
71059cf4267d80d2934b85a7046fc7b64900378b

We don't expect to see tx1 in mempool, right? It's not being rebroadcast, as it's inactive?

```suggestion
        assert tx1_txid not in self.nodes[0].getrawmempool()
```
Might be good to have a check for this behavior.

ishaanam replied:
I've added an assertion that the tx isn't in the mempool.

glozow commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +130,165 @@ def run_test(self):
         assert_equal(former_conflicted["confirmations"], 1)
         assert_equal(former_conflicted["blockheight"], 217)
 
+    def test_mempool_conflict(self):
```
Not sure if in scope for this PR, but it would be nice to have a test for how the wallet sees descendants of conflicting transactions (e.g. if Bob creates a child off of tx1 or tx2).

ishaanam replied:
Yes, I will add these in a follow-up PR.

ishaanam replied:
Actually, I have just added a test in this PR.

glozow reviewed (commented):
Approach ACK

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1386,6 +1385,36 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+
+    const uint256& conflict_hash = tx->GetHash();
+
+    for (const CTxIn& tx_in : tx->vin) {
```
With this PR, if a transaction's parent is replaced, then it should become `TxStateMempoolConflicted`. If the conflict is removed, then it becomes inactive.

glozow replied:
Ah makes sense! I think it would be good to add a test for this case?

ishaanam replied:
I have now added a test for descendants of conflicted transactions.

ryanofsky commented on src/wallet/transaction.h:
```diff
@@ -48,7 +48,12 @@ struct TxStateBlockConflicted {
+    std::string toString() const { return strprintf("BlockConflicted (block=%s, height=%i)", conflicting_block_hash.ToString(), conflicting_block_height); }
+};
+
+//! State of transaction with one or more mempool conflicts.
```
In commit "wallet: track mempool conflicts" (a6ae5b23b1497ab6f4899db49348db623700a2d8)

I think it would be useful to add that mempool conflicted transactions are serialized as inactive transactions, so the distinction between TxStateMempoolConflicted and TxStateInactive only exists at runtime.

ishaanam replied:
I have added a comment indicating this.

ryanofsky reviewed (commented):
Code review 71059cf4267d80d2934b85a7046fc7b64900378b

I'm starting to review this, but I don't understand how it interacts with abandoned transactions. Is it still possible to abandon a transaction with conflicts with the mempool? If a transaction is already abandoned and we know it conflicts with the mempool does `isMempoolConflicted` still return false?

It seems to me like a more straightforward way to implement this feature would be to add a `bool mempool_conflicts` field to `TxStateInactive` rather than adding a new `TxStateMempoolConflicted` state. The description of `TxStateInactive` says "May conflict with the mempool, or with an unknown block, or be abandoned, never broadcast, or rejected from the mempool for another reason." So the inactive state was meant to handle transactions that could have conflicts or be inactive for a variety of reasons.

I think I would recommend removing the new `MempoolConflicted` state here, but otherwise keeping this PR unchanged. The mempool tracking feature, renames, and other cleanups here do seem very nice, so it is good to see this work.

ishaanam committed:
test: Add tests for wallet mempool conflicts

ishaanam commented on src/wallet/transaction.h:
```diff
@@ -258,6 +265,8 @@ class CWalletTx
     CTransactionRef tx;
     TxState m_state;
 
+    std::set<uint256> mempool_conflicts;
```
Done

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1419,6 +1419,36 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
         RefreshMempoolStatus(it->second, chain());
     }
+
+    const uint256& conflict_hash = tx->GetHash();
```
Done

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1456,6 +1488,39 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
+                    wtx.mempool_conflicts.erase(conflict_txid);
+
+                    if (wtx.mempool_conflicts.empty()) {
+                        // If this wtx has no other mempool conflicts, erase this wtx entry
```
In 5fbf18b89d6768bed1be95dcbc98797df5788dfc "wallet: track mempool conflicts"

I'm a little bit confused by the "erase this wtx entry" part of the comment. AFAICT, no entries are being erased, Perhaps this should say:

```suggestion
                        // If this wtx has no other mempool conflicts, mark it as inactive
```

ishaanam replied:
Good point, this comment is from a previous version of the PR. I have fixed it.

achow101 commented:
> It seems to me like a more straightforward way to implement this feature would be to add a `bool mempool_conflicts` field to `TxStateInactive` rather than adding a new `TxStateMempoolConflicted` state. The description of `TxStateInactive` says "May conflict with the mempool, or with an unknown block, or be abandoned, never broadcast, or rejected from the mempool for another reason." So the inactive state was meant to handle transactions that could have conflicts or be inactive for a variety of reasons.

Personally, I don't like assigning further meaning to `TxStateInactive`. It's kind of a catch-all "transaction is neither in a block nor in the mempool, and we don't know why" (except for abandoned, and even then, I don't like how abandoned is a part of it and not a separate state). This PR introduces a distinct state of "transaction currently has things in the mempool that conflict with it" which is specific and we know the reason that something is not in the mempool, so I think it makes sense to be a separate state. I think it's more useful to carve out from `TxStateInactive` more specific states with reasons that we know for sure why something is in that state.

ryanofsky commented:
> Personally, I don't like how `TxStateInactive` is so broad. [...]

Sure, but then I'm pretty sure we need to add a `bool abandoned;` field to the `TxStateBlockConflicted` struct so we don't lose track of whether the user has abandoned the transaction.

Alternately, we could go more in the other direction, getting rid of the inactive catch-all and dividing TxStateInactive into 4 states:

- TxStateAbandonedAndConflictsWithMempool
- TxStateAbandonedAndProbablyDoesntConflictWithMempool
- TxStateNotAbandonedAndConflictsWithMempool
- TxStateNotAbandonedAndProbablyDoesntConflictWithMempool

Personally I think just keeping the `TxStateInactive` state with `bool abandoned; bool mempool_conflicts;` is better than subdividing it, because most wallet code should just need to consider whether transactions are active or not, without reverse-engineering the causes of transaction becoming inactive and making special cases out of them. But any of these approaches would be fine as long as we aren't losing track of whether a transaction is abandoned after a conflict is detected.

ryanofsky commented:
> Personally, I don't like how `TxStateInactive` is so broad. [...]

I will also add that I think dividing the TxStateInactive state instead of just adding a new field to it makes this PR considerably harder to review because you have to manually search for every line of code that is accessing the `m_state` field directly or indirectly and verify that it is either:

1. Treating the TxStateInactive and TxStateMempoolConflicted states equivalently and not changing behavior
2. Treating the TxStateInactive and TxStateMempoolConflicted states differently and changing behavior in a useful way.

If can avoid adding a new top-level state it would make this PR easier to review, because you could focus on the second case, and easily see where behavior is changing and not changing, which I don't think is possible with the current approach.

This is not a fundamental objection, but might be something to consider.

achow101 commented:
> Sure, but then I'm pretty sure we need to add a `bool abandoned;` field to the `TxStateMempoolConflicted` struct so we don't lose track of whether the user has abandoned the transaction.

We already lose track of that if a conflict is created and that conflict confirms. Abandoned is not a permanent state - as soon as it shows back up in the mempool, it'll no longer be abandoned. As soon as a conflict appears in a block, it'll no longer be abandoned. Likewise, if a transaction shows up in the mempool that conflicts with it, it should reflect that it is now conflicted and no longer needs to be marked as abandoned. Abandoned has always been a transient state since someone else could rebroadcast the transaction to the network. Once there's something else that makes the transaction invalid, I don't think it's useful to keep track of whether it is abandoned, so we wouldn't need to have more states to track that.

ryanofsky commented:
> Abandoned is not a permanent state - as soon as it shows back up in the mempool, it'll no longer be abandoned.

It makes sense that a transaction in the mempool can't be treated as abandoned. But I don't see why the abandoned state be should be cleared just because a mempool heuristic happened to detect a conflict. That just seems like a bad UX and a confusion of two orthogonal concepts. The abandoned state is transient in the sense that it can be overriden by mempool events and block confirmations, but it should be less transient than the mempool conflicted state because it (1) reflects user intent and (2) is actually serialized to disk, so it doesn't seem to me something that should be thrown away unnecessarily.

Also, the PR description doesn't mention anything about discarding the abandoned state so this seems to me like an unintentional bug than a sensible design choice.

I would just recommend adding the conflict tracking feature without adding a new top level state and without changing the behavior of abandoned transactions (especially unintentionally). It would shrink this PR, make it safer and easier to review, and provide all the benefits of the current implementation. IMO, if we really do want to change the transaction state diagram and add new top level states, I think it would be better to do in a separate PR not tied to adding a new tracking feature.

ishaanam commented:
>  Is it still possible to abandon a transaction with conflicts with the mempool?

Yes, because of [this](https://github.com/bitcoin/bitcoin/pull/27307/commits/3de89342dcd44bdde4027a91cb27793dc0231847#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8L1313-R1313) change. 

> If a transaction is already abandoned and we know it conflicts with the mempool does isMempoolConflicted still return false?

Yes

> Sure, but then I'm pretty sure we need to add a bool abandoned; field to the TxStateMempoolConflicted struct so we don't lose track of whether the user has abandoned the transaction.

If the transaction has been abandoned, then it won't be marked as `TxStateMempoolConflicted` because of [these](https://github.com/bitcoin/bitcoin/pull/27307/commits/3de89342dcd44bdde4027a91cb27793dc0231847#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8R1443-R1444) lines of code. Although the conflicting transactions are kept track of regardless, the actual transaction state only changes if the transaction is inactive (not abandoned).

> I would just recommend adding the conflict tracking feature without adding a new top level state and without changing the behavior of abandoned transactions (especially unintentionally). It would shrink this PR, make it safer and easier to review, and provide all the benefits of the current implementation. IMO, if we really do want to change the transaction state diagram and add new top level states, I think it would be better to do in a separate PR not tied to adding a new tracking feature.

From what I can tell, this PR is not changing the behavior of abandoned transactions. That being said, if anybody still thinks that `TxStateMempoolConflicted` should be removed in favor of just adding a boolean to `TxStateInactive`, I would be willing to implement that. I just think that `TxStateMempoolConflicted` makes more of a transition towards having more specific transaction states, which seems to be something we want.

furszy reviewed (commented):
light review, left few comments, nothing big.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/27307#issuecomment-1926350798

Thanks ishaanam for pointing out the lines of code that answer my questions. Since original links got broken in rebase, here are stable links that point to the two lines of code which prioritize the abandoned state over the conflicted state:

- https://github.com/bitcoin/bitcoin/commit/3de89342dcd44bdde4027a91cb27793dc0231847#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8L1313-R1313
- https://github.com/bitcoin/bitcoin/commit/3de89342dcd44bdde4027a91cb27793dc0231847#diff-1f2db0e4d5c12d109c7f0962333c245b49b696cb39ff432da048e9d6c08944d8R1443-R1444

This is better than what I was afraid of, since the abandoned state consistently takes priority over the mempoolconflicted state. But this behavior is fragile and unnecessarily complicates our internal model and potential RPC and GUI interfaces, because there's no logical reason for the abandoned and mempool conflicted states to contradict each other, when both can be true at the same time.

If we take the current approach in 3471ebc8c6ee268ad501525121ea3f3a221fb5ad, we will have to be careful going forward, whenever we do any state change, to remember that the abandoned state is supposed to take priority over the mempool conflicted state (this is not even documented anywhere), and decide how we want to deal with potentially odd states like `mempool_conflicts` set being nonempty when the transaction is not in the `MempoolConflicted` state.

For background, I already think it's bad that abandoned and block-conflicted states exclude each other, and that the block conflicted state takes priority over the abandoned state and resets it. But the block conflicted state is pretty permanent, so it's basically just a UI quirk if a block conflict causes the abandoned state to be forgotten about. There are also backwards compatibility benefits to storing the block conflict and abandoned bit exclusively in the same field.

In the current case, though, there are no backwards compatibility concerns, so it seems wrong and unnecessary to go out our way to create an unenforced, undocumented, and confusing hierarchy of states: _MempoolConflicted < Abandoned < BlockConflicted_, when there is a simpler alternative of just straightforwardly representing the information we have about transactions directly.

It seems pretty clear to me that it would make this change simpler, safer and easier to review, make the code more straightforward and maintainable long term, and avoid potential bugs and confusion in the UI if it just represented abandoned and mempool conflicted states independently and didn't try to establish a hierarchy between them.

I think a good approach would avoid adding a `TxStateMempoolConflicted` top-level state, and just add a `bool mempool_conflicts` or `std::set<Txid> mempool_conflicts` member to the `TxStateInactive` struct. And it is good to rename `TxStateConflicted` to `TxStateBlockConflicted` too for clarity. If you're skeptical of this approach, I'd be curious to know why, but we can also experiment in a separate PR if you are concerned there are benefits in this PR which would be lost.

If it would be helpful, I could make another PR next week to compare the two approaches. But I'd be curious to know what you think. There is a lot of a complexity and a long history of bugs in wallet transaction state tracking (https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking) so I think think it is worth taking a little more time to get this right.

ishaanam commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +131,253 @@ def run_test(self):
+        assert_equal(alice.listunspent(), [])
+        assert_equal(alice.getbalance(), 0)
+
+        bob.sendall([self.nodes[2].getnewaddress()])
+        self.generate(self.nodes[2], 1)
```
Yes, I've added a comment for that.

ishaanam commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +131,253 @@ def run_test(self):
+
+        # Check that the 0th unspent is now available because the transaction spending it has been replaced in the mempool
+        assert_equal(alice.listunspent(), [unspents[0]])
+        assert_equal(alice.getbalance(), 25)
```
I have added a lot more comments to the test which should clarify this.

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1456,6 +1488,39 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
+                            return TxUpdate::CHANGED;
+                        }
+                    }
+                }
```
Done

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1419,6 +1419,38 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+        for (TxSpends::const_iterator _it = range.first; _it != range.second; ++_it) {
+            // if this tx is the same as the tx which was just added to the mempool, continue
+            auto it = mapWallet.find(_it->second);
+            if (it->second.tx->GetWitnessHash() == conflict_wtxid) continue;
```
> Would be good to explain that you are doing the mapWallet lookup because the wtx could have been updated.

I just realized that the wtx could not have been updated, because for it to be updated, `transactionAddedToMempool` would need to be called, which would be after this call, right? So I have reverted the change which compares witness hashes. 

> Also, as you already know that the output belongs to one of the wallet's txs, could:

Yes, I have implemented this.

ishaanam commented:
> I think a good approach would avoid adding a TxStateMempoolConflicted top-level state, and just add a bool mempool_conflicts or std::set<Txid> mempool_conflicts member to the TxStateInactive struct.

Thanks for your feedback, I have implemented this.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1420,6 +1420,37 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+                // Add this conflict to the wtx's entry (create one if one does not already exist)
+                wtx.mempool_conflicts.insert(conflict_txid);
+
+                if (wtx.isInactive()) {
+                    wtx.m_state = TxStateInactive{wtx.isAbandoned(), /*mempool_conflicted=*/ true};
```
In commit "wallet: track mempool conflicts" (2c867784b2a719c9e701e19e389c3adfa7d7faae)

Can drop this code if `TxStateInactive::mempool_conflicted` member is dropped as suggested above.

ishaanam replied:
Done

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1457,6 +1488,36 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
+
+            auto try_updating_state = [&](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
+                // Remove the previously conflicting transaction from this wtx's entry
+                wtx.mempool_conflicts.erase(conflict_txid);
+
+                if (wtx.mempool_conflicts.empty()) {
+                    // If this wtx has no other mempool conflicts, it is now considered inactive
+                    if (wtx.isMempoolConflicted()) {
+                        wtx.m_state = TxStateInactive{};
+                        return TxUpdate::CHANGED;
+                    }
+                }
+                return TxUpdate::UNCHANGED;
```
In commit "wallet: track mempool conflicts" (2c867784b2a719c9e701e19e389c3adfa7d7faae)

This has the same problems as the new transactionAddedToMempool code above. Would suggest avoiding the need for state transitions, and always updating the mempool_conflicts set recursively for BlockConflicted as well as Inactive transactions:

```c++
if (!wtx.mempool_conflicts.erase(conflict_txid)) return TxUpdate::UNCHANGED;
return TxUpdate::CHANGED;
```

ishaanam replied:
Done

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1420,6 +1420,37 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+    const Txid& conflict_txid = tx->GetHash();
+
+    for (const CTxIn& tx_in : tx->vin) {
+        // If the wallet contains no transactions that spend the same prevout, continue
+        if (mapTxSpends.count(tx_in.prevout) == 0) continue;
+
+        std::pair<TxSpends::const_iterator, TxSpends::const_iterator> range;
+        range = mapTxSpends.equal_range(tx_in.prevout);
+
+        // For each wallet transaction spending this prevout...
+        for (TxSpends::const_iterator _it = range.first; _it != range.second; ++_it) {
```
In commit "wallet: track mempool conflicts" (2c867784b2a719c9e701e19e389c3adfa7d7faae)

Just IMO, but

```c++
auto range =  mapTxSpends.equal_range(tx_in.prevout);
for (auto it = range.first; it != range.second; ++it)
```

would make core more readable, dropping the long type declarations and extra underscore. But feel free to keep current style if that is your preference.

ishaanam replied:
I have implemented most of this, except for removing the extra underscore, because `it` was already defined above.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1420,6 +1420,37 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+
+    for (const CTxIn& tx_in : tx->vin) {
+        // If the wallet contains no transactions that spend the same prevout, continue
+        if (mapTxSpends.count(tx_in.prevout) == 0) continue;
```
In commit "wallet: track mempool conflicts" (2c867784b2a719c9e701e19e389c3adfa7d7faae)

There's no reason for this line. If the count is 0, equal_range will return an empty range and the result will be the same.

ishaanam replied:
Fixed here and for `transactionRemovedFromMempool`.

ryanofsky commented on test/functional/wallet_conflicts.py:
```diff
@@ -28,6 +29,20 @@ def get_utxo_of_value(self, from_tx_id, search_value):
 
     def run_test(self):
+        """
+        The following tests check the behavior of the wallet when
```
In commit "test: Add tests for wallet mempool conflicts" (c891a65067f57ba87e443d0cdc4f7e6b434c436f)

Ignore this suggestion if it requires too much work, but IMO it would be really helpful for review if as much of the new test code as possible was added in the first commit before the C++ code changes, and the tests were updated alongside the code changes in subsequent commits.

Independently of this PR, the fact that a commit like 2c867784b2a719c9e701e19e389c3adfa7d7faae can make a number of behavior changes and not need any significant test updates shows that we are lacking a lot of test coverage for these code paths presently. So it would be very reassuring to add the test coverage before making the changes and see the changes having the expected effects on tests, instead of making the changes blindly and adding the test coverage later.

ishaanam replied:
I have made the first commit the commit with tests, and they are updated to reflect the changes in the following commits.

ryanofsky reviewed (commented):
Code review c891a65067f57ba87e443d0cdc4f7e6b434c436f.

Thanks for the updates! A lot of nice changes and test coverage here, and I think the PR looks close to ready to ACK, though I did leave a number of suggestions below. Most of these should be very simple or simplify code, so should not create too much work.

ishaanam commented on src/wallet/transaction.h:
```diff
@@ -335,7 +335,7 @@ class CWalletTx
 
     bool isAbandoned() const { return state<TxStateInactive>() && state<TxStateInactive>()->abandoned; }
-    bool isConflicted() const { return state<TxStateConflicted>(); }
+    bool isConflicted() const { return state<TxStateBlockConflicted>(); }
```
Done

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -752,8 +752,10 @@ bool CWallet::IsSpent(const COutPoint& outpoint) const
-            int depth = GetTxDepthInMainChain(mit->second);
-            if (depth > 0  || (depth == 0 && !mit->second.isAbandoned()))
+            // An output is considered spent if a spending transaction
+            // is either confirmed, or in the mempool, or inactive (but not abandoned)
```
Done

ishaanam commented on src/wallet/transaction.h:
```diff
@@ -258,6 +262,10 @@ class CWalletTx
 
+    // Transactions that conflict directly with the tx, or
+    // that conflict with an ancestor transaction
+    std::set<Txid> mempool_conflicts;
```
Done

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -755,7 +755,7 @@ bool CWallet::IsSpent(const COutPoint& outpoint) const
             // is either confirmed, or in the mempool, or inactive (but not abandoned)
             const auto& wtx = mit->second;
-            if (wtx.isConfirmed() || wtx.InMempool() || (wtx.isInactive() && !wtx.isAbandoned()))
+            if (wtx.isConfirmed() || wtx.InMempool() || (wtx.isInactive() && !wtx.isAbandoned() && !wtx.isMempoolConflicted()))
```
The preceding was more of a refactor, while this commit aims to have `IsSpent` return false for txos spent by mempool-conflicted transactions.

ishaanam commented on src/wallet/transaction.h:
```diff
@@ -335,7 +343,9 @@ class CWalletTx
 
     bool isAbandoned() const { return state<TxStateInactive>() && state<TxStateInactive>()->abandoned; }
-    bool isConflicted() const { return state<TxStateBlockConflicted>(); }
+    bool isConflicted() const { return isBlockConflicted() || isMempoolConflicted(); }
```
I have updated the commit message.

ishaanam commented on src/wallet/transaction.h:
```diff
@@ -57,9 +57,13 @@ struct TxStateBlockConflicted {
+      : abandoned(abandoned),
+        mempool_conflicted(mempool_conflicted)
+        {}
+    std::string toString() const { return strprintf("Inactive (abandoned=%i, mempool_conflicted=%i)", abandoned, mempool_conflicted); }
```
I have removed the `mempool_conflicted` variable.

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1420,6 +1420,37 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+                    wtx.m_state = TxStateInactive{wtx.isAbandoned(), /*mempool_conflicted=*/ true};
+                    return TxUpdate::CHANGED;
+                } else {
+                    return TxUpdate::UNCHANGED;
```
Good point, this bug should be fixed now.

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1420,6 +1420,37 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+                }
+            };
+
+            RecursiveUpdateTxState(_it->second, try_updating_state);
```
Done.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1423,6 +1427,27 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+            if (wallet_tx.tx->GetHash() == conflict_txid) continue;
+
+            auto try_updating_state = [&](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
+                wtx.mempool_conflicts.insert(conflict_txid);
+
+                return TxUpdate::CHANGED;
```
In commit "wallet: track mempool conflicts" (a650caab9f11be7f5927f9aa556bb7d2a8ebed33)

Would be nice to add a check that conflict txid is actually inserted as assumed:

```diff
-                wtx.mempool_conflicts.insert(conflict_txid);
-
+                Assume(wtx.mempool_conflicts.insert(conflict_txid).second);
```

ishaanam replied:
It seems like that is not always true, so I have changed this code to do something similar to `transactionRemovedFromMempool` where it only returns `TxUpdate::CHANGED` if the txid is actually inserted.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1460,6 +1485,25 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
+        for (auto _it = range.first; _it != range.second; ++_it) {
+
+            auto try_updating_state = [&](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
+                // Remove the previously conflicting transaction from this wtx's entry
+                if (wtx.mempool_conflicts.erase(conflict_txid)) return TxUpdate::CHANGED;
+
+                return TxUpdate::UNCHANGED;
```
In commit "wallet: track mempool conflicts" (a650caab9f11be7f5927f9aa556bb7d2a8ebed33)

Unimportant style point, but usually it makes more sense to return early when state is *not* changing than when it is changing:

```diff
-                if (wtx.mempool_conflicts.erase(conflict_txid)) return TxUpdate::CHANGED;
-
-                return TxUpdate::UNCHANGED;
+                if (!wtx.mempool_conflicts.erase(conflict_txid)) return TxUpdate::UNCHANGED;
+                return TxUpdate::CHANGED;
```

So if someone wants to add extra code logging or dealing with the state change, there is a place to insert it after the early return.

Alternately this could be simplified to

```c++
return wtx.mempool_conflicts.erase(conflict_txid)) ? TxUpdate::CHANGED : TxUpdate::UNCHANGED;
```

ishaanam replied:
Done

ryanofsky commented on test/functional/wallet_conflicts.py:
```diff
@@ -346,7 +346,6 @@ def test_descendants_with_mempool_conflicts(self):
         tx1_child = bob.signrawtransactionwithwallet(raw_tx)['hex']
         tx1_child_txid = bob.sendrawtransaction(tx1_child)
 
-        # Currently neither tx1 nor tx1_child should have any conflicts
```
In commit "wallet: track mempool conflicts" (a650caab9f11be7f5927f9aa556bb7d2a8ebed33)

Should this comment be restored? Unclear why it is being removed here

ishaanam replied:
That was a mistake, should be fixed now.

ryanofsky reviewed (approved):
Code review ACK 923734ef6528e5d4e13748db7387f44d02237d4d. Seems like a very nice change that improves the wallet balance calculation and ability to spend, and also returns useful information about the mempool.

Thanks for the updates, too. I think comparing current a650caab9f11be7f5927f9aa556bb7d2a8ebed33 to previous a6ae5b23b1497ab6f4899db49348db623700a2d8, the resulting code is simpler and the charges are more direct and easier to review.

I left some more suggestions below, but most are not important and none are critical so feel free to ignore them.

ishaanam commented on src/wallet/transaction.h:
```diff
@@ -258,6 +258,13 @@ class CWalletTx
+    // directly with the transaction, or that conflict
+    // with an ancestor transaction. This set is empty
+    // if the transaction is confirmed, but can be
+    // non-empty in other states.
```
Done

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1423,6 +1427,27 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+        for (auto _it = range.first; _it != range.second; ++_it) {
+            // if this tx is the same as the tx which was just added to the mempool, continue
+            CWalletTx& wallet_tx = mapWallet.at(_it->second);
+            if (wallet_tx.tx->GetHash() == conflict_txid) continue;
```
Done

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -758,7 +758,7 @@ bool CWallet::IsSpent(const COutPoint& outpoint) const
             // transactions in known, potentially active states.
             const auto& wtx = mit->second;
-            if (wtx.isConfirmed() || wtx.InMempool() || (wtx.isInactive() && !wtx.isAbandoned()))
+            if (!wtx.isAbandoned() && !wtx.isMempoolConflicted() && !wtx.isBlockConflicted())
```
In commit "wallet: track mempool conflicts" (9fa7f813281d8520a5eb5adbd29fe3a0d2ab5a4f)

Would suggest deleting the "An output is considered spent..." comment above now that the code is simpler and it no longer adds any new information. The part of the comment about checking for the "inverse condition" is also no longer true, because the code is now checking directly if wtx is conflicted or abandoned, instead of checking for the inverse condition (that it is confirmed, or in mempool, or inactive and not abandoned).

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -752,8 +752,13 @@ bool CWallet::IsSpent(const COutPoint& outpoint) const
+            // check for the inverse condition and only consider spending
+            // transactions in known, potentially active states.
+            const auto& wtx = mit->second;
+            if (wtx.isConfirmed() || wtx.InMempool() || (wtx.isInactive() && !wtx.isAbandoned()))
```
In commit "wallet: use CWalletTx member functions to determine tx state" (92b3204b69e5ec25cbb29bd158f25d2b07e48cf6)

Would suggest simplifying this code and deleting the comment so this is just:

```c++
const auto& wtx = mit->second;
if (!wtx.isAbandoned() && !wtx.isBlockConflicted())
```

Using the simpler condition would make this commit easier to understand, and make the change in the next commit more obvious where the line becomes:

```c++
if (!wtx.isAbandoned() && !wtx.isMempoolConflicted() && !wtx.isBlockConflicted())
```

ryanofsky reviewed (approved):
Code review ACK b853688c92d3745d5ec2ec5119584148084e7c1c. Just more simplifications since that last review. This PR keeps getting smaller and simpler, and seems like a very obvious improvement now.

I left another suggestion (really 2 related suggestions), but it is not very important. This looks good in its current form.

furszy commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +138,249 @@ def run_test(self):
+        bob.sendrawtransaction(raw_tx2)
+        assert_equal(bob.getbalances()["mine"]["untrusted_pending"], 0)
+
+        bob.sendrawtransaction(tx1_conflict_conflict) # kick tx1_conflict out of the mempool
+        bob.sendrawtransaction(raw_tx1) #re-broadcast tx1 because it is no longer conflicted
```
In d3f7764c802fe:

nit: 
Could move the `tx1_conflict_conflict` creation here. Basically because it adds unneeded information at the beginning of the test that reviewers/readers have to carry over. The tx only exist to kick `tx1_conflict` out of the mempool so `tx1` can be re-accepted at this point.

furszy commented on test/functional/wallet_conflicts.py:
```diff
@@ -123,5 +138,249 @@ def run_test(self):
+        self.sync_mempools()
+        self.generate(self.nodes[2], 1)
+
+    def test_descendants_with_mempool_conflicts(self):
```
In d3f7764c802fe:
Could add wallet balance checks along this test.

furszy reviewed (commented):
left few small things. Will get deeper.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1363,8 +1363,12 @@ void CWallet::MarkConflicted(const uint256& hashBlock, int conflicting_height, c
+    RecursiveUpdateTxState(&batch, tx_hash, try_updating_state);
+}
+
+void CWallet::RecursiveUpdateTxState(WalletBatch* batch, const uint256& tx_hash, const TryUpdatingStateFn& try_updating_state) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
+    // Do not flush the wallet here for performance reasons
```
re: https://github.com/bitcoin/bitcoin/pull/27307#discussion_r1516106925

> This comment relates to the second argument of the `WalletBatch` constructor. Need to move it to the caller side, it does not mean anything inside `RecursiveUpdateTxState`.

Right the comment makes sense where it was previously above `WalletBatch`, and shouldn't be moved

ryanofsky reviewed (approved):
Code review ACK dbcc4a75e4a1108f4a04c22cfae363848994c24d. Changes since last review were a few test improvements and some more simplifications. This looks very good. Left suggestions below, all minor, none important. This PR seems like a clear improvement and should be merged if it gets more ACKs.

ishaanam commented on src/wallet/interfaces.cpp:
```diff
@@ -101,7 +101,7 @@ WalletTxStatus MakeWalletTxStatus(const CWallet& wallet, const CWalletTx& wtx)
     result.is_abandoned = wtx.isAbandoned();
     result.is_coinbase = wtx.IsCoinBase();
-    result.is_in_main_chain = wallet.IsTxInMainChain(wtx);
+    result.is_in_main_chain = wtx.isConfirmed();
```
Done

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1418,6 +1421,20 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+
+    for (const CTxIn& tx_in : tx->vin) {
+        // Mark all already spent inputs, and all txs that consumes them, as conflicted
+        for (auto range = mapTxSpends.equal_range(tx_in.prevout); range.first != range.second; range.first++) {
```
Done

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1455,6 +1472,35 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
+            const uint256& spent_id = range.first->second;
+
+            RecursiveUpdateTxState(/*batch=*/nullptr, spent_id, [&conflict_txid](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
+                return !wtx.mempool_conflicts.erase(conflict_txid) ? TxUpdate::UNCHANGED : TxUpdate::CHANGED;
```
Done

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1455,6 +1472,35 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
+            RecursiveUpdateTxState(nullptr, _it->second, try_updating_state);
+        }
+        */
+
```
Done

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1418,6 +1421,20 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
         RefreshMempoolStatus(it->second, chain());
     }
+
+    const Txid& conflict_txid = tx->GetHash();
```
Done

achow101 commented:
ACK 5952292133d6cc889f51ae771f2e0557311e1efe

ryanofsky reviewed (approved):
Code review ACK 5952292133d6cc889f51ae771f2e0557311e1efe. Just small suggested changes since last review

furszy commented on src/wallet/rpc/transactions.cpp:
```diff
@@ -417,6 +421,10 @@ static std::vector<RPCResult> TransactionDescriptionString()
            }},
            {RPCResult::Type::STR_HEX, "replaced_by_txid", /*optional=*/true, "Only if 'category' is 'send'. The txid if this tx was replaced."},
            {RPCResult::Type::STR_HEX, "replaces_txid", /*optional=*/true, "Only if 'category' is 'send'. The txid if this tx replaces another."},
+           {RPCResult::Type::ARR, "mempoolconflicts", "Transactions that directly conflict with either this transaction or an ancestor transaction",
```
Not for this PR just so you don't have to re-touch it but it would be good to describe the difference between `walletconflicts` and `mempoolconflicts` inside the help text. `walletconflicts` description is quite vague.

furszy reviewed (commented):
ACK 59522921
